### ISSUE TITLE
Registry map doesn't work on air-gapped environments

### ISSUE THREAD

**Issue by joaojacome** [GoogleContainerTools/kaniko#3182]
**Actual behavior**
Whenever pulling images from a mapped registry, Kaniko is trying to access the original registry. This makes the build on air-gapped environments fail.
```
2.199 INFO[0001] Retrieving image manifest alpine:latest      
2.199 INFO[0001] Retrieving image alpine:latest from mapped registry my-registry.io 
2.201 WARN[0001] Failed to retrieve image alpine:latest from remapped registry my-registry.io: unable to complete operation after 0 attempts, last error: Get "https://index.docker.io/v2/": dial tcp 127.0.0.1:443: connect: connection refused. Will try with the next registry, or fallback to the original registry. 
2.201 INFO[0001] Retrieving image alpine:latest from registry index.docker.io 
2.201 error building image: unable to complete operation after 0 attempts, last error: Get "https://index.docker.io/v2/": dial tcp 127.0.0.1:443: connect: connection refused
```

**Expected behavior**

I'd expect the original registry not to be called.

**To Reproduce**
1. Create a Dockerfile with

    ```
    FROM gcr.io/kaniko-project/executor:debug as kaniko
    RUN echo "FROM alpine:latest" >> /workspace/Dockerfile
    RUN /kaniko/executor --registry-map "index.docker.io=my-registry.io/docker.io" --destination my-registry.io
    ```

1. Build it with `docker build . --add-host=index.docker.io:127.0.0.1`

**Additional Information**
 - Kaniko Image (fully qualified with digest)
 `gcr.io/kaniko-project/executor@sha256:9eb8f2aff512f360d2775c80de462bf9d49599f0749f745af8f5af5f480cb4b9`
 
 **Triage Notes for the Maintainers**
 

 
 | **Description** | **Yes/No** |
 |----------------|---------------|
 | Please check if this a new feature you are proposing        | <ul><li>- [ ] </li></ul>|
 | Please check if the build works in docker but not in kaniko | <ul><li>- [ ] </li></ul>| 
 | Please check if this error is seen when you use `--cache` flag | <ul><li>- [ ] </li></ul>|
 | Please check if your dockerfile is a multistage dockerfile | <ul><li>- [ ] </li></ul>|

**Comment by joaojacome** [comment]
This change here fixed it for me, but it doesn't look good, and it was done in a trial-and-error basis. I'm not familiar with the code base (or golang), so I'd refrain myself from submitting a PR.

```diff
diff --git a/pkg/image/remote/remote.go b/pkg/image/remote/remote.go
index 787c5381..7ac8fe13 100644
--- a/pkg/image/remote/remote.go
+++ b/pkg/image/remote/remote.go
@@ -57,11 +57,13 @@ func RetrieveRemoteImage(image string, opts config.RegistryOptions, customPlatfo
 			regToMapTo, repositoryPrefix := parseRegistryMapping(registryMapping)
 
 			insecurePull := opts.InsecurePull || opts.InsecureRegistries.Contains(regToMapTo)
+			ref, err := name.ParseReference(repositoryPrefix+image, name.WeakValidation)
 
 			remappedRepository, err := remapRepository(ref.Context(), regToMapTo, repositoryPrefix, insecurePull)
 			if err != nil {
 				return nil, err
 			}
+			remappedRepository.Registry, err = name.NewRegistry(regToMapTo, name.WeakValidation, name.Insecure)
 
 			remappedRef := setNewRepository(ref, remappedRepository)
 
```

**Comment by joaojacome** [comment]
I found out this has been caused by having `docker.io` or `gcr.io` as part of the mapped registry URL.
This doesn't work:

`/kaniko/executor --registry-map "index.docker.io=my-registry.io/docker.io" --destination my-registry.io/my-image:latest`

This works:

`/kaniko/executor --registry-map "index.docker.io=my-registry.io/dockerio" --destination my-registry.io/my-image:latest`

**Comment by next-jesusmanuelnavarro** [comment]
> I found out this has been caused by having `docker.io` or `gcr.io` as part of the mapped registry URL. This doesn't work:
> 
> `/kaniko/executor --registry-map "index.docker.io=my-registry.io/docker.io" --destination my-registry.io/my-image:latest`
> 
> This works:
> 
> `/kaniko/executor --registry-map "index.docker.io=my-registry.io/dockerio" --destination my-registry.io/my-image:latest`

What do you exactly mean?  Did you **just** drop the dots from your registry-map or did you create a new mirror without the dots and then used the new mirror with registry-map?

**Comment by joaojacome** [comment]
I created a new mirror without the dots

**Comment by next-jesusmanuelnavarro** [comment]
> I created a new mirror without the dots

Yes, it makes sense.  I stepped onto this same problem and it seems in fact it is the dots.

1. Trying to mirror to a repository with dots:
